forge/src/main/java/me/lucko/luckperms/forge/mixin/core/server/network/ServerLoginPacketListenerImplMixin.java

```diff
@@ -0,0 +1,100 @@
+ *  in the Software without restriction, including without limitation the rights
+ *  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ *  copies of the Software, and to permit persons to whom the Software is
+ *  furnished to do so, subject to the following conditions:
+ *
+ *  The above copyright notice and this permission notice shall be included in all
+ *  copies or substantial portions of the Software.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ *  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ *  SOFTWARE.
+ */
+
+package me.lucko.luckperms.forge.mixin.core.server.network;
+
+import com.mojang.authlib.GameProfile;
+import me.lucko.luckperms.forge.event.ConnectionEvent;
+import net.minecraft.network.Connection;
+import net.minecraft.server.network.ServerLoginPacketListenerImpl;
+import net.minecraftforge.common.MinecraftForge;
+import org.spongepowered.asm.mixin.Final;
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.Shadow;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+
+import javax.annotation.Nullable;
+
+/**
+ * Mixin into {@link ServerLoginPacketListenerImpl} for posting {@link ConnectionEvent}
+ */
+@Mixin(value = ServerLoginPacketListenerImpl.class)
+public abstract class ServerLoginPacketListenerImplMixin {
+
+    @Shadow
+    @Final
+    public Connection connection;
+
+    @Shadow
+    @Nullable
+    GameProfile gameProfile;
+
+    @Shadow
+    protected abstract GameProfile createFakeProfile(GameProfile p_10039_);
+
+    private boolean luckperms$authenticated;
+
+    @Inject(
+            method = "tick",
+            at = @At(
+                    value = "INVOKE",
+                    target = "Lnet/minecraftforge/network/NetworkHooks;tickNegotiation(Lnet/minecraft/server/network/ServerLoginPacketListenerImpl;Lnet/minecraft/network/Connection;Lnet/minecraft/server/level/ServerPlayer;)Z",
+                    remap = false
+            )
+    )
+    private void onTickNegotiation(CallbackInfo callbackInfo) {
+        if (this.gameProfile == null) {
+            return;
+        }
+
+        if (!this.gameProfile.isComplete()) {
+            this.gameProfile = this.createFakeProfile(this.gameProfile);
+        }
+
+        if (this.luckperms$authenticated) {
+            return;
+        }
+
+        this.luckperms$authenticated = true;
+        MinecraftForge.EVENT_BUS.post(new ConnectionEvent.Auth(this.connection, this.gameProfile));
```

Comment:
Just to confirm - we can safely perform blocking calls here without disrupting the server thread?

Reply:
This would run on the server thread however the only usage for this event is to allow LuckPerms to setup a `CompletableFuture` for asynchronously preloading player data while the Forge negotiation takes place, there is no blocking operation.

On `PlayerEvent.LoadFromFile` the `CompletableFuture` is blocked until completion, This might actually be an issue although the solution would to replace `get()` with `getNow(false)` so that throws if it's not completed.

I could remove these events entirely however there would need to be a statically accessible instance of LPForgePlugin as it's not possible to obtain it.

https://github.com/LuckPerms/LuckPerms/blob/ab07f1f5391534a199ba38e195f6df9c7263e04f/forge/src/main/java/me/lucko/luckperms/forge/listeners/ForgeConnectionListener.java#L59-L94
https://github.com/LuckPerms/LuckPerms/blob/ab07f1f5391534a199ba38e195f6df9c7263e04f/forge/src/main/java/me/lucko/luckperms/forge/listeners/ForgeConnectionListener.java#L106-L111

Reply:
Hmm `getNow(false)` isn't going to work.

We need to block the player connection handling while the database load is performed. This could take >100ms, so it's really important that we block and wait, and not just "hope" that it completes async by the time we need it.

All other platforms have an event or some sort of construct for this: 

https://github.com/LuckPerms/LuckPerms/blob/c35958b88ec05022270461d2b8c522c8d991a008/bukkit/src/main/java/me/lucko/luckperms/bukkit/listeners/BukkitConnectionListener.java#L91

https://github.com/LuckPerms/LuckPerms/blob/c35958b88ec05022270461d2b8c522c8d991a008/fabric/src/main/java/me/lucko/luckperms/fabric/listeners/FabricConnectionListener.java#L80-L81

You can see the Fabric impl with Mixins here: https://github.com/FabricMC/fabric/tree/1.18.2/fabric-networking-api-v1

https://github.com/FabricMC/fabric/blob/1.18.2/fabric-networking-api-v1/src/main/java/net/fabricmc/fabric/mixin/networking/ServerLoginNetworkHandlerMixin.java#L59-L65
https://github.com/FabricMC/fabric/blob/1.18.2/fabric-networking-api-v1/src/main/java/net/fabricmc/fabric/impl/networking/server/ServerLoginNetworkAddon.java#L70-L111

My strong preference would be to request + wait for this functionality to exist in Forge. I'm relucant to hack our own hooks in with Mixins for this sort of thing...

Reply:
Hey @lucko, I was wondering if you could provide your thoughts and reasoning as to why LuckPerms needs this event added to Forge and why consumers of LuckPerms need the user data early.
https://github.com/MinecraftForge/MinecraftForge/pull/8599